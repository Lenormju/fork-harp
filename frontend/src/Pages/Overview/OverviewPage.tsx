import { Page } from "Components/Page"
import { OnQuerySuccess } from "Components/Utilities/OnQuerySuccess"
import { useOverviewDataQuery } from "Domain/Overview"
import { useSystemSettingsQuery } from "Domain/System"

import { TransactionsOverviewChart } from "./Components/OverviewChart/TransactionsOverviewChart"
import { TransactionsOverview } from "./OverviewCharts"

export const OverviewPage = () => {
  const query = useOverviewDataQuery()
  const settingsQuery = useSystemSettingsQuery()
  interface ProxyData {
    endpoints?: { name: string; port: number; url: string; description?: string }[]
  }

  return (
    <Page title="Overview" description="Useful insights">
      <OnQuerySuccess query={query}>
        {(query) => {
          return <TransactionsOverviewChart data={query.data} title="Transactions Overview" className="mb-10" />
        }}
      </OnQuerySuccess>
      <OnQuerySuccess query={settingsQuery}>
        {(query) => {
          const proxyData = query.data.proxy as ProxyData
          const endpointsNames = proxyData.endpoints?.map((endpoint: { name: string }) => endpoint.name)
          return (
            <div className="grid grid-cols-2 gap-4">
              {endpointsNames &&
                endpointsNames?.length > 1 &&
                endpointsNames.map((endpoint: string, index: number) => (
                  <TransactionsOverview key={index} endpoint={endpoint} title={endpoint} className="border p-2" />
                ))}
            </div>
          )
        }}
      </OnQuerySuccess>
    </Page>
  )
}
