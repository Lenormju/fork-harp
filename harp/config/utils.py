import importlib.util
from typing import TYPE_CHECKING, Type

from .defaults import DEFAULT_NAMESPACES

if TYPE_CHECKING:
    from .applications import Application


def get_application_class_name(name):
    return "".join(map(lambda x: x.title().replace("_", ""), name.rsplit(".", 1)[-1:])) + "Application"


def resolve_application_name(spec):
    if "." not in spec:
        for namespace in DEFAULT_NAMESPACES:
            _candidate = ".".join((namespace, spec))
            if importlib.util.find_spec(_candidate):
                return _candidate
    if importlib.util.find_spec(spec):
        return spec

    raise ModuleNotFoundError(f"No application named {spec}.")


#: Cache for application types
_application_types = {}


def get_application_type(name: str) -> Type["Application"]:
    """
    Returns the application class for the given application name.

    :param name:
    :return:
    """
    if name not in _application_types:
        application_spec = importlib.util.find_spec(name)
        if not application_spec:
            raise ValueError(f'Unable to find application "{name}".')

        try:
            application_module = __import__(".".join((application_spec.name, "__app__")), fromlist=["*"])
        except ModuleNotFoundError as exc:
            raise ModuleNotFoundError(
                f'A python package for application "{name}" was found but it is not a valid harp application. '
                'Did you forget to add an "__app__.py"?'
            ) from exc

        application_class_name = get_application_class_name(name)
        if not hasattr(application_module, application_class_name):
            raise AttributeError(f'Application module for {name} does not contain a "{application_class_name}" class.')

        _application_types[application_spec.name] = getattr(application_module, application_class_name)
        _application_types[application_spec.name].name = application_spec.name

    if name not in _application_types:
        raise RuntimeError(f'Unable to load application "{name}", application class definition not found.')

    return _application_types[name]
