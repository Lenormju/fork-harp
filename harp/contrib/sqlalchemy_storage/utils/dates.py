from sqlalchemy import func
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql.expression import ColumnElement
from sqlalchemy.types import DateTime


# Code from https://stackoverflow.com/questions/51662547/truncate-hour-day-week-month-year-in-sqlalchemy
class TruncDatetime(ColumnElement):
    type = DateTime()

    inherit_cache = True

    def __init__(self, precision, expr):
        self.precision = precision
        self.expr = expr

    @property
    def _from_objects(self):
        return self.expr._from_objects


@compiles(TruncDatetime, "postgresql")
def compile_trunc_postgresql(element, compiler, **kw):
    return compiler.process(func.date_trunc(element.precision, element.expr))


_modifiers = {
    "year": ("start of year",),
    "month": ("start of month",),
    # This does not account for locale specific first day of week. 1 day
    # is added so that the 1st day of week won't truncate to previous week.
    # Replace 'weekday 0' with 'weekday 1', if you'd like first day of
    # week to be Monday (in accordance with ISO 8601)
    "week": ("1 day", "weekday 0", "-7 days", "start of day"),
    "day": ("start of day",),
}


@compiles(TruncDatetime, "sqlite")
def compile_trunc_sqlite(element, compiler, **kw):
    try:
        precision = element.precision.effective_value
    except AttributeError:
        precision = element.precision
    expr = element.expr
    modifiers = _modifiers.get(precision)

    if modifiers:
        return compiler.process(func.datetime(expr, *modifiers))

    if precision == "hour":
        return compiler.process(
            func.datetime(expr, func.strftime("-%M minutes", expr), func.strftime("-%f seconds", expr))
        )

    if precision == "minute":
        return compiler.process(func.datetime(expr, func.strftime("-%f seconds", expr)))

    if precision == "second":
        return compiler.process(
            func.datetime(expr, func.strftime("-%f seconds", expr), func.strftime("%S seconds", expr))
        )

    raise NotImplementedError(f"Truncating {precision} is not supported for SQLite")
