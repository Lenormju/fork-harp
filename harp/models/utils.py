def generate_transaction_id_ksuid():
    from ksuid import KsuidMs

    return str(KsuidMs())


def generate_transaction_id_nanoid():
    from nanoid import generate
    from nanoid.resources import alphabet

    return generate(alphabet=alphabet[2:])


def generate_transaction_id_ulid():
    from ulid import ULID

    return str(ULID())
