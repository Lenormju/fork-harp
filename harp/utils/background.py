import asyncio
from inspect import iscoroutinefunction

from harp import get_logger

logger = get_logger(__name__)


class AsyncWorkerQueue:
    def __init__(self):
        self._queue = asyncio.Queue()
        self._task = asyncio.create_task(self())
        self._running = True

    async def __call__(self):
        while True:
            try:
                item, ignore_errors = await self._queue.get()
            except RuntimeError:
                # queue is closed
                break
            try:
                await item()
            except Exception as e:
                if not ignore_errors:
                    logger.exception(f"Error while executing queued task: {e}")
            finally:
                self._queue.task_done()

    async def push(self, item, /, *, ignore_errors=False):
        if not self._running:
            raise RuntimeError("Queue is closed.")
        if not iscoroutinefunction(item):
            raise ValueError(f"Unknown item type: {type(item)}, expecting coroutine function.")

        await self._queue.put((item, ignore_errors))

    async def wait_until_empty(self):
        await self._queue.join()

    async def close(self):
        self._running = False
        await self.wait_until_empty()
