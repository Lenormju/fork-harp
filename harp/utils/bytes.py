def ensure_bytes(x):
    if isinstance(x, bytes):
        return x
    return bytes(x, "utf-8")
