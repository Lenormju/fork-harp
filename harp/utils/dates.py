from datetime import date, datetime
from typing import Optional


def ensure_date(x) -> Optional[date]:
    if x is None:
        return None
    if isinstance(x, datetime):
        return x.date()
    if isinstance(x, date):
        return x
    return datetime.strptime(x, "%Y-%m-%d").date()
