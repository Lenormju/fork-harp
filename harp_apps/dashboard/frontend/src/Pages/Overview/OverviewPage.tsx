import { useState } from "react"
import { Sparklines, SparklinesCurve, SparklinesLine, SparklinesSpots } from "react-sparklines"
import tw, { styled } from "twin.macro"

import ApdexBadge from "Components/Badges/ApdexBadge.tsx"
import { StyledJumboBadge } from "Components/Badges/StyledJumboBadge.tsx"
import { Page, PageTitle } from "Components/Page"
import { OnQuerySuccess } from "Components/Utilities/OnQuerySuccess"
import { useSystemSettingsQuery } from "Domain/System"
import { ButtonGroup } from "ui/Components/ButtonGroup"
import { Pane } from "ui/Components/Pane"
import { H2 } from "ui/Components/Typography"

import { TransactionsOverview } from "./Containers"

import { useSummaryDataQuery } from "../../Domain/Overview/useSummaryDataQuery.tsx"

export const OverviewPage = () => {
  const settingsQuery = useSystemSettingsQuery()

  const summaryQuery = useSummaryDataQuery()

  interface ProxyData {
    endpoints?: { name: string; port: number; url: string; description?: string }[]
  }
  const [timeRange, setTimeRange] = useState("7d")
  const setCurrentTimeRange = (timeRange: string) => {
    setTimeRange(timeRange)
  }

  const buttonProps = [
    { key: "1h", title: "1h" },
    { key: "24h", title: "24h" },
    { key: "7d", title: "7d" },
    { key: "1m", title: "1m" },
    { key: "1y", title: "1y" },
  ]

  const StyledSummaryPane = styled(Pane)(() => [tw`overflow-hidden h-32 xl:h-40 2xl:h-48 relative`])

  return (
    <Page title={<PageTitle title="Overview" />}>
      <div className="grid grid-cols-3 gap-4">
        <StyledSummaryPane>
          <H2>Performances</H2>
          <OnQuerySuccess query={summaryQuery}>
            {(query) => {
              return (
                <>
                  <div className="flex self-center relative z-10">
                    <ApdexBadge score={query.data.apdex.mean} size="xl" className="ring-1 ring-white" />
                  </div>
                  <Sparklines
                    min={0}
                    data={query.data.apdex.data.map((d: { value: number }) => d.value)}
                    style={{ margin: "-3px", position: "absolute", bottom: 0, left: 0, zIndex: 0 }}
                  >
                    <SparklinesCurve color="#ADD8E6" />
                    <SparklinesSpots />
                  </Sparklines>
                </>
              )
            }}
          </OnQuerySuccess>
        </StyledSummaryPane>
        <StyledSummaryPane>
          <H2>Transactions</H2>
          <OnQuerySuccess query={summaryQuery}>
            {(query) => {
              return (
                <>
                  <div className="flex self-center relative z-10">
                    <StyledJumboBadge className="bg-white ring-1 ring-blue-200" size="xl" color="black">
                      {query.data.transactions.rate}{" "}
                      <span className="font-light">/{query.data.transactions.period}</span>
                    </StyledJumboBadge>
                  </div>
                  <Sparklines
                    min={0}
                    data={query.data.transactions.data.map((d: { value: number }) => d.value)}
                    style={{ margin: "-3px", position: "absolute", bottom: 0, left: 0 }}
                  >
                    <SparklinesLine color="#ADD8E6" />
                    <SparklinesSpots />
                  </Sparklines>
                </>
              )
            }}
          </OnQuerySuccess>
        </StyledSummaryPane>
        <StyledSummaryPane>
          <H2>Errors</H2>
          <OnQuerySuccess query={summaryQuery}>
            {(query) => {
              return (
                <>
                  <div className="flex self-center relative z-10">
                    <StyledJumboBadge className="bg-white ring-1 ring-red-200" size="xl" color="black">
                      {query.data.errors.rate} <span className="font-light">/{query.data.errors.period}</span>
                    </StyledJumboBadge>
                  </div>
                  <Sparklines
                    min={0}
                    data={query.data.errors.data.map((d: { value: number }) => d.value)}
                    style={{ margin: "-3px", position: "absolute", bottom: 0, left: 0 }}
                  >
                    <SparklinesLine color="#e6b3ad" />
                    <SparklinesSpots />
                  </Sparklines>
                </>
              )
            }}
          </OnQuerySuccess>
        </StyledSummaryPane>
      </div>

      <div className="flex justify-end my-2">
        <ButtonGroup buttonProps={buttonProps} current={timeRange} setCurrent={setCurrentTimeRange} />
      </div>

      <Pane className="mb-4">
        <TransactionsOverview title="Transactions Overview" timeRange={timeRange} />
      </Pane>

      <OnQuerySuccess query={settingsQuery}>
        {(query) => {
          const proxyData = query.data.proxy as ProxyData
          const endpointsNames = proxyData.endpoints?.map((endpoint: { name: string }) => endpoint.name)
          return (
            <div className="grid grid-cols-2 gap-4">
              {endpointsNames &&
                endpointsNames?.length > 1 &&
                endpointsNames.map((endpoint: string, index: number) => (
                  <Pane key={index}>
                    <TransactionsOverview endpoint={endpoint} title={endpoint} timeRange={timeRange} />
                  </Pane>
                ))}
            </div>
          )
        }}
      </OnQuerySuccess>
    </Page>
  )
}
