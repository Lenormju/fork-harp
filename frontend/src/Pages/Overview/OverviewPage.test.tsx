import { renderWithClient } from "test/utils"
import { expect, it } from "vitest"

import { OverviewPage } from "./OverviewPage"

it("renders the title and data when the query is successful", async () => {
  const result = renderWithClient(<OverviewPage />)

  await result.findByText("endpoint1")
  expect(result.container).toMatchSnapshot()
})
