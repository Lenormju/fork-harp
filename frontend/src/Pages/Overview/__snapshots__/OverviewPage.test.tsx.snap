// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the title and data when the query is successful 1`] = `
<div>
  <main>
    <div
      class="mt-4 mb-4"
    >
      <h1
        class="css-9lrcch"
      >
        Overview
      </h1>
      <p
        class="css-11bp3eh"
      >
        Useful insights
      </p>
    </div>
    <div
      class="mt-4"
    >
      <div
        class="sm:-mx-6 lg:-mx-8"
      >
        <section
          class="inline-block min-w-full max-w-full py-2 align-middle sm:px-6 lg:px-8"
        >
          <div
            class="flex justify-end my-2"
          >
            <span
              class="isolate inline-flex rounded-md shadow-sm"
            >
              <button
                class="css-1gwafep"
                type="button"
              >
                1h
              </button>
              <button
                class="css-1gwafep"
                type="button"
              >
                24h
              </button>
              <button
                class="css-v7wx26"
                type="button"
              >
                7d
              </button>
              <button
                class="css-1gwafep"
                type="button"
              >
                1m
              </button>
              <button
                class="css-1gwafep"
                type="button"
              >
                1y
              </button>
            </span>
          </div>
          <div
            class="mb-4 css-14upriz"
          >
            <div>
              <h2
                class="css-13yts7e"
              >
                Transactions Overview
              </h2>
              <div
                style="display: flex; align-items: center;"
              >
                <div
                  class="flex flex-col items-center"
                >
                  <div
                    class="flex self-center text-3xl"
                  >
                    <span
                      class="px-1 py-0.5 text-white font-semibold bg-emerald-400"
                    >
                      A+
                    </span>
                  </div>
                  <div
                    class="grid grid-cols-2 text-xs text-left align-text-bottom items-center self-center ml-10 mt-10"
                  >
                    <span
                      class="font-bold"
                    >
                      Mean duration:
                    </span>
                    <span>
                      0.03
                       s
                    </span>
                    <span
                      class="font-bold"
                    >
                      Errors:
                    </span>
                    <span>
                      50
                      %
                    </span>
                  </div>
                </div>
                <div
                  class="recharts-responsive-container"
                  style="width: 90%; height: 300px; min-width: 0;"
                />
              </div>
            </div>
          </div>
          <div
            class="grid grid-cols-2 gap-4"
          >
            <div
              class="css-14upriz"
            >
              <div>
                <h2
                  class="css-13yts7e"
                >
                  endpoint1
                </h2>
                <div
                  style="display: flex; align-items: center;"
                >
                  <div
                    class="flex flex-col items-center"
                  >
                    <div
                      class="flex self-center text-3xl"
                    >
                      <span
                        class="px-1 py-0.5 text-white font-semibold bg-emerald-400"
                      >
                        A+
                      </span>
                    </div>
                    <div
                      class="grid grid-cols-2 text-xs text-left align-text-bottom items-center self-center ml-10 mt-10"
                    >
                      <span
                        class="font-bold"
                      >
                        Mean duration:
                      </span>
                      <span>
                        0.03
                         s
                      </span>
                      <span
                        class="font-bold"
                      >
                        Errors:
                      </span>
                      <span>
                        50
                        %
                      </span>
                    </div>
                  </div>
                  <div
                    class="recharts-responsive-container"
                    style="width: 90%; height: 300px; min-width: 0;"
                  />
                </div>
              </div>
            </div>
            <div
              class="css-14upriz"
            >
              <div>
                <h2
                  class="css-13yts7e"
                >
                  endpoint2
                </h2>
                <div
                  style="display: flex; align-items: center;"
                >
                  <div
                    class="flex flex-col items-center"
                  >
                    <div
                      class="flex self-center text-3xl"
                    >
                      <span
                        class="px-1 py-0.5 text-white font-semibold bg-emerald-400"
                      >
                        A+
                      </span>
                    </div>
                    <div
                      class="grid grid-cols-2 text-xs text-left align-text-bottom items-center self-center ml-10 mt-10"
                    >
                      <span
                        class="font-bold"
                      >
                        Mean duration:
                      </span>
                      <span>
                        0.03
                         s
                      </span>
                      <span
                        class="font-bold"
                      >
                        Errors:
                      </span>
                      <span>
                        50
                        %
                      </span>
                    </div>
                  </div>
                  <div
                    class="recharts-responsive-container"
                    style="width: 90%; height: 300px; min-width: 0;"
                  />
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </main>
</div>
`;
